sys/viyml

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python3
+import errno
+import fcntl
+import signal
+from argparse import ArgumentParser
+from contextlib import contextmanager
+
+import yaml
+from ocflib.misc.shell import edit_file
+
+TIMEOUT = 2
+
+
+# taken from https://stackoverflow.com/questions/5255220/fcntl-flock-how-to-implement-a-timeout
+@contextmanager
+def timeout(seconds):
+    original_handler = signal.signal(signal.SIGALRM, lambda *_: None)
+
+    try:
+        signal.alarm(seconds)
+        yield
+    finally:
+        signal.alarm(0)
+        signal.signal(signal.SIGALRM, original_handler)
+
+
+def edit_yaml(target_file):
+    content = target_file.read()
+    while True:
+        content = edit_file(content)
+        try:
+            yaml.load(content)
+            # passed the yaml check, now write at the start
+            # of the file and truncate the old content
+            target_file.seek(0)
+            target_file.write(content)
+            target_file.truncate()
+            break
+        except yaml.YAMLError:
+            pass
+
+        idunno = input('Syntax Error: Try again? [y/n] ')
+        idunno = idunno.strip().lower()
+        if idunno != 'y' and idunno != 'yes':
+            break
+
+
+def main():
+    parser = ArgumentParser()
+    parser.add_argument('path', nargs=1)
+    args = parser.parse_args()
+
+    path = args.path[0]
+
+    with open(path, 'r+') as target_file:
+        with timeout(TIMEOUT):
```

Comment:
Why the timeout? Unless we anticipate frequent situations in which the file is locked now, but will be unlocked within 2 seconds, I don't see this as being particularly useful. (Or is it NFS weirdness?)

Reply:
It's more that if, say, I'm modifying the file, and you open the file, your terminal won't hang until you either `SIGINT` or I quit out from the file. I'm not attached to the timeout though